Source/WebKit/UIProcess/mac/WebViewImpl.mm

```diff
@@ -5111,6 +5127,113 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
     m_page->setCompositionAsync(text, WTFMove(underlines), WTFMove(highlights), selectedRange, replacementRange);
 }
 
+#if HAVE(INLINE_PREDICTIONS)
+bool WebViewImpl::allowsInlinePredictions() const
+{
+    return m_page->preferences().inlinePredictionsEnabled() && [NSSpellChecker respondsToSelector:@selector(isAutomaticInlineCompletionEnabled)] && NSSpellChecker.isAutomaticInlineCompletionEnabled;
+}
+
+void WebViewImpl::showInlinePredictionsForCandidate(NSTextCheckingResult *candidate, NSRange absoluteSelectedRange, NSRange oldRelativeSelectedRange)
+{
+    if (!candidate)
+        return;
+
+    auto postLayoutData = postLayoutDataForContentEditable();
+    if (!postLayoutData)
+        return;
+
+    if (m_lastStringForCandidateRequest != postLayoutData->stringForCandidateRequest)
+        return;
+
+    NSRange relativeSelectedRange = NSMakeRange(postLayoutData->candidateRequestStartPosition, postLayoutData->selectedTextLength);
+    if (absoluteSelectedRange.location < relativeSelectedRange.location || absoluteSelectedRange.length != relativeSelectedRange.length)
+        return;
+
+    // Make sure the selected range didn’t change while we were making an asynchronous call to get it.
+    if (!NSEqualRanges(oldRelativeSelectedRange, relativeSelectedRange))
+        return;
+
+    NSString *paragraphContextForCandidateRequest = postLayoutData->paragraphContextForCandidateRequest;
+    auto offsetSelectionRect = postLayoutData->selectionBoundingRect;
+    offsetSelectionRect.move(0, offsetSelectionRect.height());
+
+    NSUInteger offset = absoluteSelectedRange.location - relativeSelectedRange.location;
+    NSTextCheckingResult *adjustedCandidate = [candidate resultByAdjustingRangesWithOffset:offset];
+
+    NSSpellChecker *spellChecker = [NSSpellChecker sharedSpellChecker];
+
+    if (![spellChecker respondsToSelector:@selector(showCompletionForCandidate:selectedRange:offset:inString:rect:view:completionHandler:)])
```

Comment:
Is this selector check necessary?

I think we also need a declaration in `NSSpellCheckerSPI.h` for OpenSource builds.